### PR TITLE
Support explain_weights(pipeline)

### PR DESCRIPTION
**Pull request by jnothman**
Towards fixing #15, allowing `explain_weights` for `Pipeline`s. Does not handle `explain_predictions`; tracking the provenance of features back to parts of text may be tricky in a convoluted pipeline/union structure.

No tests yet.

Introduces new `eli5.transform.transform_feature_names` singledispatch, which could do with more implementations.

I think the exact forms of the names should remain experimental... or indeed we could refuse to provide default implementations for anything but vectorizers, selectors, `Pipeline` and `FeatureUnion`, letting users register their own.

Rubbish example:
```python
from sklearn import *
from eli5 import explain_weights
from eli5.sklearn.transform import register_experimental_feature_names
from IPython.display import display
register_experimental_feature_names()
est = pipeline.make_pipeline(pipeline.FeatureUnion([('words', pipeline.make_pipeline(feature_extraction.text.CountVectorizer(),
                                                                                     feature_extraction.text.TfidfTransformer())),
                                                    ('chars', feature_extraction.text.CountVectorizer(analyzer='char'))]),
                             feature_selection.SelectKBest(feature_selection.chi2, k=10),
                             linear_model.LogisticRegression())
bunch = datasets.fetch_20newsgroups(categories=['alt.atheism', 'talk.religion.misc', 'rec.autos'])
est.fit(bunch.filenames, bunch.target)

display(explain_weights(est, target_names=bunch.target_names))
```

I'm not sure I'll be able to complete this patch any time soon, though I am keen to have something like this merged! Help is welcome.

### CODE REVIEW

**Comment by jnothman** [comment]
[Note: @kmike's wish that we only calculate feature names when we've determined that they are important requires on knowing dependencies between input and output features for each transformer, i.e. something like https://github.com/scikit-learn/enhancement_proposals/pull/5]

File: eli5/sklearn/explain_weights.py [hunk 1]
```diff
@@ -412,3 +414,13 @@ def _features(target_id):
             method='linear model',
             is_regression=True,
         )
+
+
+@explain_weights.register(Pipeline)
+def explain_weights_pipeline(estimator, feature_names=None, **kwargs):
+    last_estimator = estimator.steps[-1][1]
+    transform_pipeline = Pipeline(estimator.steps[:-1])
+    feature_names = transform_feature_names(transform_pipeline, feature_names)
```
Comment (kmike) [inline eli5/sklearn/explain_weights.py]: In other cases `feature_names` argument of explain_weights has higher priority than feature names extracted from estimator, i.e. it allows to override feature names completely. The way `feature_names` are used here looks fine though. But we should document this gotcha.

File: eli5/sklearn/explain_weights.py [hunk 2]
```diff
@@ -412,3 +414,13 @@ def _features(target_id):
...
+    out = explain_weights(last_estimator, feature_names=feature_names)
```
Comment (kmike) [inline eli5/sklearn/explain_weights.py]: it makes sense to pass `**kwargs` here

File: eli5/sklearn/transform.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+from sklearn.preprocessing import Imputer
+
+from eli5.transform import transform_feature_names
+
+
+@transform_feature_names.register(Pipeline)
+def _pipeline_names(est, in_names=None):
+    names = in_names
+    for name, trans in est.steps:
+        if trans is not None:
```
Comment (kmike) [inline eli5/sklearn/transform.py]: when can `trans` be None?

File: eli5/transform.py [hunk 4]
```diff
@@ -0,0 +1,11 @@
+"""Handling transformation pipelines in explanations"""
+
+from singledispatch import singledispatch
+
+
+@singledispatch
+def transform_feature_names(transformer, in_names=None):
+    if hasattr(transformer, 'get_feature_names'):
+        return transformer.get_feature_names()
```
Comment (kmike) [inline eli5/transform.py]: ❤️

**Comment by kmike** [comment]
Thanks @jnothman! If you don't have time we can pick up the development. I've left some comments in order to not forget about them.

File: eli5/sklearn/transform.py (hunk 3 above)
Comment (jnothman) [inline eli5/sklearn/transform.py]: Changes released in scikit-learn 0.18 (scikit-learn/scikit-learn#1769) allow parts of a pipeline or union to be replaced or disabled in a grid search.

File: eli5/sklearn/explain_weights.py (hunk 2 above)
Comment (jnothman) [inline eli5/sklearn/explain_weights.py]: Yes, I'd intended to :)
Fixed

File: eli5/sklearn/explain_weights.py (hunk 1 above)
Comment (jnothman) [inline eli5/sklearn/explain_weights.py]: The passed feature names (sort of) takes priority here and is merely transformed, except where there is a vectoriser in the pipeline. But I think I should be using eli5's get_feature_names here to allow a vectoriser here. Is that correct?

**Comment by jnothman** [comment]
How much do you think we should prefabricate feature name designs and how much leave it to the user? As I said, we could import implementations of transform_feature_names for Pipeline, FeatureUnion, SelectorMixin by default, and leave others unimplemented; or provide default implementations in some optional import like eli5.sklearn.experimental. the issue is that sometimes a user will  to see "topic1" for brevity, but a list of the top features on the topic would be more helpful!

**Comment by jnothman** [comment]
It may similarly be useful to show "scale(featurename)" but some users would rather the scale transform elided. There are many ways to name features.

Which of course leads to the possibility of feature descriptions, but I think that can wait for later

**Comment by jnothman** [comment]
More untested code pushed :)

**Comment by jnothman** [comment]
I now see that it should be trivial to support the same feature in `explain_prediction`

File: eli5/sklearn/transform.py (hunk 3 above)
Comment (kmike) [inline eli5/sklearn/transform.py]: Thanks for the explanation!

**Comment by kmike** [comment]
> How much do you think we should prefabricate feature name designs and how much leave it to the user? As I said, we could import implementations of transform_feature_names for Pipeline, FeatureUnion, SelectorMixin by default, and leave others unimplemented; or provide default implementations in some optional import like eli5.sklearn.experimental. the issue is that sometimes a user will to see "topic1" for brevity, but a list of the top features on the topic would be more helpful!

I think providing as much default feature names as possible is a good thing. As for topics, what do you think about showing only a few top features? `eli5.formatters.html` could allow to expand topics on a client, or just show more words in `title` attribute.

**Comment by kmike** [comment]
> It may similarly be useful to show "scale(featurename)" but some users would rather the scale transform elided. There are many ways to name features.
Which of course leads to the possibility of feature descriptions, but I think that can wait for later

Good points. 
I think providing defaults like `scale(featurename)` is fine; customizability can be added / documented later if there is demand.

File: eli5/sklearn/transform.py [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+
+    transform_feature_names.register(Binarizer)(
+        transform_feature_names_binarizer)  # Perhaps should also be identity
+    transform_feature_names.register(PolynomialFeatures)(
+        transform_feature_names_polynomial_features)
+
+    # TODO: OneHotEncoder. scikit-learn#6441 doesn't appear complete
+
+    transform_feature_names.register(TfidfTransformer)(
+        make_tfn_featurewise('tfidf', lambda est: len(est.idf_)))
```
Comment (kmike) [inline eli5/sklearn/transform.py]: make_tfn_featurewise has only 1 argument

Comment (kmike) [inline eli5/sklearn/transform.py]: This + https://github.com/TeamHG-Memex/eli5/commit/724c4232ba753e4ada84f16a0755d74a3e9a5750 should make mypy env pass. I haven't figured out how to make a pull request to your pull request :)

File: eli5/sklearn/transform.py [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+    def transform_names(est, in_names=None):
+        weights = orig_weights = getattr(est, weight_attr)
+        in_names = _get_feature_names(est, feature_names=in_names,
+                                      num_features=weights.shape[1])
+        if max_features:
+            # TODO: support sparse weights
+            if abs:
+                weights = np.abs(weights)
+            order = np.argsort(weights, axis=1)[::-1]
+            top_names = np.take(in_names, order[:, :max_features])
```
Comment (kmike) [inline eli5/sklearn/transform.py]: probably not too important, but it seems this could be made faster using [eli5.utils.argsort_k_largest](https://github.com/TeamHG-Memex/eli5/blob/e2f2acfeac5f7a13f13b314e486069660287d7b0/eli5/utils.py#L6)

Comment (jnothman) [inline eli5/sklearn/transform.py]: Curious. Have you benchmarked `argsort_k_largest`? Perhaps I'm just used to supporting old versions of numpy where `argpartition` requires `argsort` anyway.

Comment (jnothman) [inline eli5/sklearn/transform.py]: (and if you want speed, use `a.take(b, mode='clip')` rather than `a[b]`)

File: eli5/sklearn/transform.py (hunk 5 above)
Comment (jnothman) [inline eli5/sklearn/transform.py]: You are able to directly edit my branch, I think. What trouble do you have in making a PR to my branch?

File: eli5/sklearn/transform.py [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+@transform_feature_names.register(SelectorMixin)
+def _select_names(est, in_names=None):
+    mask = est.get_support(indices=False)
+    in_names = _get_feature_names(est, feature_names=in_names,
+                                  num_features=len(mask))
+    return [in_names[i] for i in np.flatnonzero(mask)]
+
+
+# XXX Should these generic things be in eli5.transform?
+# I've left them here to make use of eli5.sklearn.utils.get_feature_names
```
Comment (jnothman) [inline eli5/sklearn/transform.py]: What do you think of this comment, @kmike? Is `eli5.sklearn.utils.get_feature_names` in the right place?

Comment (kmike) [inline eli5/sklearn/transform.py]: There are two kinds of "scikit-learningness": things which work with objects from scikit-learn and things which work with objects which follow scikit-learn API ideas. `get_feature_names` (which is used in xgboost as well) is currently in eli5.sklearn because of the second meaning.

File: eli5/sklearn/transform.py (hunk 6 above)
Comment (kmike) [inline eli5/sklearn/transform.py]: mm, I believe there was a use case where it mattered, but now I can't recall it; for small-medium examples argsort works just as fine.

**Comment by jnothman** [comment]
@kmike what do you propose to move this forward? (I've noticed not a great deal of eli5 work lately in general.) Are we happy with the API and I should find time to write tests?

**Comment by kmike** [comment]
@jnothman I'd love to have it moved forward :) To merge it we need tests. I like the API; transformer coverage is great, and we can always improve coverage in future. 

I'd also register everything by default, without marking it as experimental and requiring user to call a register function. This way explain_weights works in more cases, and if users want to override the way features are formatted it is the same amount of work (register their own function).

Other than that it looks fine.

File: eli5/sklearn/explain_weights.py [hunk 8]
```diff
@@ -412,3 +414,19 @@ def _features(target_id):
             method='linear model',
             is_regression=True,
         )
+
+
+@explain_weights.register(Pipeline)
+def explain_weights_pipeline(estimator, feature_names=None, **kwargs):
+    last_estimator = estimator.steps[-1][1]
+    transform_pipeline = Pipeline(estimator.steps[:-1])
+    if hasattr(feature_names, 'get_feature_names'):
```
Comment (kmike) [inline eli5/sklearn/explain_weights.py]: What is this condition for? I'd expect feature_names to be a list or FeatureNames instance.

**Comment by jnothman** [comment]
i think that might be for an old conception. I'm slowly working through
tests and docs, but there's a lot to test. starting bottom up.

**Comment by kmike** [comment]
@jnothman thanks for the update! What do you think about merging it partially, e.g. only the API part, Pipeline, FeatureUnion, and maybe a couple of basic transformers, to make it easier to test & document?

**Comment by jnothman** [comment]
might be sensible. Either way I don't think I'll have anything for the next
week

File: eli5/sklearn/explain_weights.py (hunk 8 above)
Comment (jnothman) [inline eli5/sklearn/explain_weights.py]: This is to reflect some of the behaviour of `eli5.sklearn.utils.get_feature_names`. If you don't care for it, I'll remove it.

Comment (jnothman) [inline eli5/sklearn/explain_weights.py]: Never mind. Cleaned up.

**Comment by jnothman** [comment]
I'll pull out a PR with just the basic (`explain_weights_pipeline` plus `transform_feature_names` defined for feature selection and pipeline).

**Comment by jnothman** [comment]
Where does further documentation belong?

**Comment by kmike** [comment]
@jnothman what further documentation are you thinking of? I think an example of how to register your own transform_feature_name handler may go to sklearn library docs.

**Comment by jnothman** [comment]
Yes, fair enough.
